File: Data Structures & Algorithms/house-robber-ii/submission-6.ts
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    rob(nums: number[]): number {
        if (nums.length == 1) { return nums[0] }

        function robArr(arr: number[]): number {
            let visited: Record<number, number> = {};

            function dfs(i: number): number {
                if (i >= arr.length) {return 0}
                if (visited[i] != null ) {
                    return visited[i]
                }
                const skip = dfs(i+1)
                const steal = arr[i]+dfs(i+2)
                visited[i] = Math.max( skip, steal)
                return visited[i]
            }
            return dfs(0)
        }

        
        return Math.max(
            robArr(nums.slice(1)),
            robArr(nums.slice(0, -1))
        )
    }
}

File: Data Structures & Algorithms/decode-ways/submission-6.swift
class Solution {
    func numDecodings(_ s: String) -> Int {
        let s = Array(s)
        guard s.count > 0 else { return 0 }

        var visited: [Int: Int] = [:]

        return dfs(0)

        func dfs(_ i: Int) -> Int {
            guard i < s.count else { return 1 }
            guard s[i] != "0" else { return 0 }
            if visited[i] != nil { return visited[i]! }

            var subSum = dfs(i+1)
            if i+1 < s.count && (s[i] == "1" || (s[i] <= "2" && s[i+1] <= "6") ) {
                subSum += dfs( i+2 )
            }
            visited[i] = subSum
            return subSum
        }
    }
}
